include/envoy/ssl/private_key/private_key.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <functional>
+#include <string>
+
+#include "envoy/api/v2/auth/cert.pb.h"
+#include "envoy/common/pure.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/ssl/private_key/private_key_callbacks.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+// Prevent a dependency loop with the forward declaration.
+class TransportSocketFactoryContext;
+} // namespace Configuration
+} // namespace Server
+
+namespace Ssl {
+
+typedef std::shared_ptr<SSL_PRIVATE_KEY_METHOD> BoringSslPrivateKeyMethodSharedPtr;
+
+class PrivateKeyOperations {
+public:
+  virtual ~PrivateKeyOperations() {}
+
+  /**
+   * Associate the private key operations instance with a SSL connection.
+   * @param ssl a SSL connection object. The BoringSSL private key API
+   * doesn't allow passing user data to the asynchronous functions as a
+   * function parameter, so this enables the private key method provider
+   * to use SSL connection custom data fields instead.
+   */
+  virtual bool associateWithSsl(SSL* ssl) PURE;
```

Comment:
Why isn't this part of `PrivateKeyOperationsCallbacks`? I.e. someone implementing the lifecycle callbacks around an async method should also be able to take an initial SSL binding call as well.

Reply:
I removed the `associateWithSsl()` function because the SSL object pointer was already available in the provider call. I also did a bunch of renaming effort which should hopefully move the naming conventions closer to the BoringSSL.